### ISSUE TITLE
Canary runs into infinite recursion loop with recursive list structure de-serialization

### ISSUE THREAD

**Issue by suushiemaniac** [pdvrieze/xmlutil#32]
Kotlin: 1.3.72 JVM
KotlinX serialization: 0.20.0
xmlutil: 0.20.0.1

When defining types than can recursively contain themselves (I'd like to avoid the discussion whether that is good design) the **de**-serialization process hangs and finally throws a `StackOverflowError` with the stacktrace jumping back and forth between `Canary.kt:208` and `Canary.kt:213` in `Canary.serialDescriptor`. Example:

```kotlin
@Serializable
data class XmlRec(val nested: List<XmlRec> = emptyList())

val recursionWonderland = XmlRec(listOf(XmlRec(listOf(XmlRec(), XmlRec())), XmlRec()))

val intoTheRabbitHole = XML.stringify(XmlRec.serializer(), recursionWonderland) // <--- this produces the expected output as defined below
val notComingBack = XML.parse(XmlRec.serializer(), intoTheRabbitHole) // <--- this call triggers the StackOverflowError
```

Albeit the arguably poor design decision, I believe that the following output representing the `recursionWonderland` variable is perfectly valid XML:

```xml
<XmlRec>
	<XmlRec>
		<XmlRec/>
		<XmlRec/>
	</XmlRec>
	<XmlRec/>
</XmlRec>
```

Just for the sake of mentioning it: I'm working with an external dictionary dump in XML format that has these kinds of design decisions and it is neither feasible (due to the immense size and historical artifacts) nor within my power to change the input format. This library is my only ray of hope for not having to work with JAXB! :smile:

**Comment by suushiemaniac** [comment]
I stand corrected: The minimal example I posted above does work. However, things still actually do break in my large-scale project (for which I cannot share the full source code at this moment in time)

Trying to come up with a better example that _actually_ breaks.

**Comment by suushiemaniac** [comment]
Nevermind, my example _does_ break as intended for the sake of this error report. You just actualy have to *de*-serialize after serialisation for the error to occur. The original post has been updated accordingly.

Sorry for the extra fuss :sweat_smile:

**Comment by pdvrieze** [comment]
I've just pushed a fix to master (f1d2b45). I haven't written a test for it, but the way it is written it should handle indirect recursion too.

**Comment by pdvrieze** [comment]
Now released in v0.20.0.10
